### PR TITLE
Enable UpgradePlan edits and edit validation

### PR DESCRIPTION
**Pull request by ipetrov117**
Suggested changes introduce the following new functionalities:
1. Users can now edit existing `UpgradePlan` resources to trigger new environment upgrades
2. Each edit is now validated by a `ValidationWebhook` that is responsible for determining when an edit can be done

Webhook validation cases include:
1. Allow edit:
   - Only if `releaseVersion` is incremented and the newly passed `releaseVersion` value is greater than the previously running `releaseVersion`
   - Only if the previous upgrade triggered by the `UpgradePlan` has finished
2. Block edit:
    - If  the newly provided `releaseVerison` is "<=" of the currently applied `releaseVersion`
    - If an edit has happened over a different field, but `releaseVersion` was not incremented
    - If there is currently a running upgrade started from this `UpgradePlan`

**Important Note: _As suggested by the kubebuilder [docs](https://book.kubebuilder.io/cronjob-tutorial/cert-manager), cert-manager is used to generate certificates that the webhook server will use. This would mean that with the introduction of these changes we will have a dependency on cert-manager from here onward._**

Testing use-cases:
1. Deployment of a new `UpgradePlan` should not be blocked by the validation webhook:
```yaml
# Successfully deployed and finished UpgradePlan
kind: UpgradePlan
metadata:
  finalizers:
  - upgradeplan.lifecycle.suse.com/finalizer
  generation: 2
  ...
  name: upgradeplan-sample
  namespace: upgrade-controller-system
  ...
spec:
  disableDrain:
    controlPlane: false
    worker: false
  releaseVersion: 3.0.1
status:
  conditions:
  - lastTransitionTime: "2024-09-02T09:03:37Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: OSUpgraded
  - lastTransitionTime: "2024-09-02T09:03:44Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: KubernetesUpgraded
  ...
  lastSuccessfulReleaseVersion: 3.0.1
  observedGeneration: 2
  sucNameSuffix: fa82540065
```
2. Edit over an existing `UpgradePlan` when the previous upgrade has finished **and** the new `releaseVersion` is greater than the old one:
```yaml
# Last successful UpgradePlan
kind: UpgradePlan
metadata:
  finalizers:
  - upgradeplan.lifecycle.suse.com/finalizer
  generation: 2
  ...
  name: upgradeplan-sample
  namespace: upgrade-controller-system
  ...
spec:
  disableDrain:
    controlPlane: false
    worker: false
  releaseVersion: 3.0.1
status:
  conditions:
  - lastTransitionTime: "2024-09-02T09:16:31Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: OSUpgraded
  - lastTransitionTime: "2024-09-02T09:17:38Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: KubernetesUpgraded
  ...
  lastSuccessfulReleaseVersion: 3.0.1
  observedGeneration: 2
  sucNameSuffix: 00f986d2f8
``` 

`kubectl edit upgradeplan upgradeplan-sample` -> `releaseVersion: 3.0.1 -> 3.0.2` 

```yaml
# Edited UpgradePlan
kind: UpgradePlan
metadata:
  finalizers:
  - upgradeplan.lifecycle.suse.com/finalizer
  generation: 3
  name: upgradeplan-sample
  namespace: upgrade-controller-system
spec:
  disableDrain:
    controlPlane: false
    worker: false
  releaseVersion: 3.0.2
status:
  conditions:
  - lastTransitionTime: "2024-09-02T09:20:50Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: OSUpgraded
  - lastTransitionTime: "2024-09-02T09:21:57Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: KubernetesUpgraded
  ...
  lastSuccessfulReleaseVersion: 3.0.2
  observedGeneration: 3
  sucNameSuffix: 2e6fba175f
```
```bash
mgmt-cluster-network:~/upgrade-controller # k get plans
NAME                                      IMAGE                                 CHANNEL   VERSION
control-plane-sl-micro-6-0-00f986d2f8     registry.suse.com/bci/bci-base:15.5             3.0.1
control-plane-sl-micro-6-0-2e6fba175f     registry.suse.com/bci/bci-base:15.5             3.0.2
control-plane-v1-28-9-rke2r1-00f986d2f8   rancher/rke2-upgrade                            v1.28.9+rke2r1
control-plane-v1-28-9-rke2r1-2e6fba175f   rancher/rke2-upgrade                            v1.28.9+rke2r1
mgmt-cluster-network:~/upgrade-controller # k get pods
NAME                                                              READY   STATUS      RESTARTS      AGE
apply-control-plane-sl-micro-6-0-00f986d2f8-on-mgmt-clust-4fvwc   0/1     Completed   0             7m41s
apply-control-plane-sl-micro-6-0-2e6fba175f-on-mgmt-clust-2dssv   0/1     Completed   0             4m21s
apply-control-plane-v1-28-9-rke2r1-00f986d2f8-on-mgmt-clu-x97m2   0/1     Completed   0             7m24s
apply-control-plane-v1-28-9-rke2r1-2e6fba175f-on-mgmt-clu-2fsxj   0/1     Completed   0             3m5s
...
mgmt-cluster-network:~/upgrade-controller # k get secrets
NAME                                         TYPE                                  DATA   AGE
...
os-upgrade-secret-sl-micro-6-0-00f986d2f8    Opaque                                1      7m45s
os-upgrade-secret-sl-micro-6-0-2e6fba175f    Opaque                                1      4m25s
...
```
3. Edit over an existing upgrade plan without changing the `releaseVersion`:
```bash
error: upgradeplans.lifecycle.suse.com "upgradeplan-sample" could not be patched: admission webhook "vupgradeplan.kb.io" denied the request: any edits over 'upgradeplan-sample' must come with an increment of the releaseVersion
You can run `kubectl replace -f /tmp/kubectl-edit-3541954845.yaml` to try this update again.
```
4. Edit over an existing upgrade plan, but setting a lower `releaseVersion` that the currently running one (`releaseVersion: 3.0.1 -> 3.0.0`):
```bash
error: upgradeplans.lifecycle.suse.com "upgradeplan-sample" could not be patched: admission webhook "vupgradeplan.kb.io" denied the request: new releaseVersion '3.0.0' must be greater than the currently applied '3.0.1' releaseVersion
You can run `kubectl replace -f /tmp/kubectl-edit-3337302911.yaml` to try this update again.
```
5. Edit over an existing upgrade plan, but setting a `releaseVersion` that does not exist (`releaseVersion: 3.0.1 -> 3.0.10`):
```bash
mgmt-cluster-network:~/upgrade-controller # k get pods
NAME                                                    READY   STATUS         RESTARTS   AGE
apply-release-manifest-3-0-10-hhkj6                     0/1     ErrImagePull   0          58s
upgrade-controller-controller-manager-bddcc7cff-5z72d   1/1     Running        0          51m
```
6. Edit over an existing upgrade plan, but when a previous upgrade is still working:
```yaml
# Plan currently in progress
kind: UpgradePlan
metadata:
  finalizers:
  - upgradeplan.lifecycle.suse.com/finalizer
  generation: 2
  ...
  name: upgradeplan-sample
  namespace: upgrade-controller-system
  ...
spec:
  disableDrain:
    controlPlane: false
    worker: false
  releaseVersion: 3.0.1
status:
  conditions:
  - lastTransitionTime: "2024-09-02T10:24:36Z"
    message: All cluster nodes are upgraded
    reason: Succeeded
    status: "True"
    type: OSUpgraded
  - lastTransitionTime: "2024-09-02T10:24:36Z"
    message: Control plane nodes are being upgraded
    reason: InProgress
    status: "False"
    type: KubernetesUpgraded
  ...
  observedGeneration: 2
  sucNameSuffix: 2de6711a59
```
```bash
error: upgradeplans.lifecycle.suse.com "upgradeplan-sample" could not be patched: admission webhook "vupgradeplan.kb.io" denied the request: detected 'InProgress' condition in upgrade plan condition list, cannot apply edit. Condition: KubernetesUpgraded
You can run `kubectl replace -f /tmp/kubectl-edit-1926073623.yaml` to try this update again.
```
7. Existing upgrade plan deletion is not blocked by the validating webhook

### CODE REVIEW

File: api/v1alpha1/upgradeplan_types.go [hunk 1]
```diff
@@ -85,8 +85,15 @@ type UpgradePlanStatus struct {
 	// +optional
 	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
 
+	// ObservedGeneration is the currently tracked generation of the UpgradePlan. Meant for internal use only.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
 	// SUCNameSuffix is the suffix added to all resources created for SUC. Meant for internal use only.
+	// Changes for each new ObservedGeneration
```
Comment (atanasdinov) [inline api/v1alpha1/upgradeplan_types.go:92]: Nit: Can we end the comment with a dot in order to have a more consistent documentation in the CRD? Also applies to `LastSuccessfulReleaseVersion`.

File: api/v1alpha1/upgradeplan_webhook.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		switch condition.Reason {
+		case UpgradeInProgress:
+			return nil, fmt.Errorf("detected 'InProgress' condition in upgrade plan condition list, cannot apply edit. Condition: %s", condition.Type)
+		case UpgradePending:
+			return nil, fmt.Errorf("detected 'Pending' condition in upgrade plan condition list, cannot apply edit. Condition: %s", condition.Type)
+		case UpgradeError:
+			// Transient error during component execution
+			return nil, fmt.Errorf("detected plan with 'Error' condition, cannot apply edit. Condition:  %s", condition.Type)
+		}
+	}
```
Comment (atanasdinov) [inline api/v1alpha1/upgradeplan_webhook.go:97]: Nit: All of these are very similar. Can we simplify using something like the following?

```
disallowingUpdateStates := []string{UpgradeInProgress, UpgradePending, UpgradeError}

for _, condition := range r.Status.Conditions {
	if slices.Contains(disallowingUpdateStates, condition.Reason) {
		return nil, fmt.Errorf("upgrade plan cannot be edited while condition '%s' is in '%s' state", condition.Type, condition.Reason)
	}
}
```

**Review by atanasdinov — APPROVED** [review]
Great work!

File: api/v1alpha1/upgradeplan_webhook.go (hunk 2 above)
Comment (ipetrov117) [reply api/v1alpha1/upgradeplan_webhook.go]: This is simpler, fixed.

**Review by atanasdinov — APPROVED** [review]

**Pull request merged**